src/controller/python/chip/yaml/YamlUtils.py

```diff
@@ -0,0 +1,130 @@
+
+    if (origin == typing.Union or origin == typing.Optional or origin == Nullable):
+        underlying_field_type = None
+
+        if field_value is NullValue:
+            for t in typing.get_args(field_type):
+                if t == Nullable:
+                    return field_value
+
+        for t in typing.get_args(field_type):
+            # Comparison below explicitly not using 'isinstance' as that doesn't do what we want.
+            if t != Nullable and t != type(None):
+                underlying_field_type = t
+                break
+
+        if (underlying_field_type is None):
+            raise ValueError(f"Can't find the underling type for {field_type}")
+
+        field_type = underlying_field_type
+
+    # Dictionary represents a data model struct.
+    if (type(field_value) is dict):
+        return_field_value = {}
+        field_descriptors = field_type.descriptor
+        for item in field_value:
+            try:
+                # We search for a matching item in the list of field descriptors
+                # for this struct and ensure we can find a field with a matching
+                # label.
+                field_descriptor = next(
+                    x for x in field_descriptors.Fields if x.Label.lower() ==
+                    item.lower())
+            except StopIteration as exc:
+                raise ValidationError(
+                    f'Did not find field "{item}" in {str(field_type)}') from None
+
+            return_field_value[field_descriptor.Label] = convert_yaml_type(
+                field_value[item], field_descriptor.Type, use_from_dict)
+        if use_from_dict:
+            return field_type.FromDict(return_field_value)
+        return return_field_value
+    elif(type(field_value) is float):
+        return float32(field_value)
+    # list represents a data model list
+    elif(type(field_value) is list):
+        list_element_type = typing.get_args(field_type)[0]
+
+        # The field type passed in is the type of the list element and not list[T].
+        for idx, item in enumerate(field_value):
+            field_value[idx] = convert_yaml_type(item, list_element_type, use_from_dict)
+        return field_value
+    # YAML conversion treats all numbers as ints. Convert to a uint type if the schema
+    # type indicates so.
+    elif (field_type == uint):
+        # Longer number are stored as strings. Need to make this conversion first.
+        value = int(field_value)
+        return field_type(value)
+    # YAML treats enums as ints. Convert to the typed enum class.
+    elif (issubclass(field_type, enum.Enum)):
+        return field_type(field_value)
+    # YAML treats bytes as strings. Convert to a byte string.
+    elif (field_type == bytes and type(field_value) != bytes):
+        if isinstance(field_value, str) and field_value.startswith(_HEX_PREFIX):
+            return bytes.fromhex(field_value[len(_HEX_PREFIX):])
+        return str.encode(field_value)
```

Comment:
`str.encode` defaults to UTF-8. This will not work for hex strings encoded with hex escapes:

```
In [18]: str.encode("\xff")
Out[18]: b'\xc3\xbf'
```

^^^ You can see that `\xff` is converted to UTF-8 unicode, not an actual octet.

I have attached a prior script I made to convert the representation from YAML to octet string. It's surprisingly annoying.

I will share a script offline.